File: python/fmvpu/amlet/alu_instruction.py
import logging
from dataclasses import dataclass
from typing import List, Tuple
from enum import IntEnum

from fmvpu.control_structures import pack_fields_to_words, unpack_words_to_fields, calculate_total_width, pack_fields_to_int


logger = logging.getLogger(__name__)


class ALUModes(IntEnum):
    NONE = 0
    ADD = 1
    ADDI = 2
    SUB = 3
    SUBI = 4
    MULT = 5
    MULT_ACC = 6
    MULT_ACC_INIT = 7
    EQ = 8
    GTE = 9
    LTE = 10
    NOT = 11
    AND = 12
    OR = 13
    RESERVED14 = 14
    RESERVED15 = 15
    SHIFT_L = 16
    SHIFT_R = 17
    RESERVED18 = 18
    RESERVED19 = 19
    RESERVED20 = 20
    RESERVED21 = 21
    RESERVED22 = 22
    RESERVED23 = 23
    RESERVED24 = 24
    RESERVED25 = 25
    RESERVED26 = 26
    RESERVED27 = 27
    RESERVED28 = 28
    RESERVED29 = 29
    RESERVED30 = 30
    JUMP = 31


@dataclass
class ALUInstruction:
    """ALU instruction for amlet"""
    mode: ALUModes = ALUModes.NONE
    src1: int = 0  # Source 1 register (d-type register)
    src2: int = 0  # Source 2 register (d-type register)
    dst: int = None   # Encoded destination register (B-register space)
    a_dst: int = None  # A-register destination (if specified)
    d_dst: int = None  # D-register destination (if specified)
    
    def __post_init__(self):
        """Set dst based on a_dst or d_dst if specified"""
        if self.mode != ALUModes.NONE:
            count = (self.a_dst is not None) + (self.d_dst is not None) + (self.dst is not None)
            if count != 1:
                raise ValueError(f"Must specifiy exactly 1 of a_dst ({self.a_dst}), d_dst({self.d_dst}) and dst({self.dst})")
    
    @classmethod
    def get_width(cls, params) -> int:
        """Get the bit width of this instruction type based on parameters"""
        return calculate_total_width(cls.get_field_specs(params))
    
    @classmethod
    def get_field_specs(cls, params) -> List[Tuple[str, int]]:
        """Get field specifications for bit packing."""
        return [
            ('mode', 5),  # 5 bits to support up to 31
            ('src1', params.d_reg_width),
            ('src2', params.d_reg_width),
            ('dst', params.b_reg_width),
        ]
    
    def encode(self, params) -> int:
        """Encode to instruction bits"""
        # Determine the actual dst value
        if self.a_dst is not None:
            # A-registers map directly to B-register space
            actual_dst = self.a_dst
        elif self.d_dst is not None:
            # D-registers map to B-register space starting at cutoff
            cutoff = max(params.n_a_regs, params.n_d_regs)
            actual_dst = cutoff + self.d_dst
        elif self.dst is not None:
            actual_dst = self.dst
        else:
            actual_dst = 0
        
        # Create a temporary object for encoding
        temp_instr = type(self)(
            mode=self.mode,
            src1=self.src1,
            src2=self.src2,
            dst=actual_dst
        )
        
        field_specs = self.get_field_specs(params)
        assert temp_instr.dst is not None
        return pack_fields_to_int(temp_instr, field_specs)
    
    @classmethod
    def from_word(cls, word: int, params) -> 'ALUInstruction':
        """Parse instruction from word"""
        field_specs = cls.get_field_specs(params)
        field_values = unpack_words_to_fields([word], field_specs, word_width=32)
        return cls(**field_values)

File: src/main/scala/zamlet/amlet/ALUInstruction.scala
package zamlet.amlet

import chisel3._
import chisel3.util._


object ALUInstr {

  object Modes extends ChiselEnum {
    val None = Value(0.U)
    val Add = Value(1.U)       
    val Addi = Value(2.U)
    val Sub = Value(3.U)
    val Subi = Value(4.U)
    val Mult = Value(5.U)
    val MultAcc = Value(6.U)
    val MultAccInit = Value(7.U)
  
    val Eq = Value(8.U)
    val Gte = Value(9.U)
    val Lte = Value(10.U)
    val Not = Value(11.U)
    val And = Value(12.U)
    val Or = Value(13.U)
    val Reserved14 = Value(14.U)
    val Reserved15 = Value(15.U)
  
    val ShiftL = Value(16.U)
    val ShiftR = Value(17.U)
    val Reserved18 = Value(18.U)
    val Reserved19 = Value(19.U)
    val Reserved20 = Value(20.U)
    val Reserved21 = Value(21.U)
    val Reserved22 = Value(22.U)
    val Reserved23 = Value(23.U)
    val Reserved24 = Value(24.U)
    val Reserved25 = Value(25.U)
    val Reserved26 = Value(26.U)
    val Reserved27 = Value(27.U)
    val Reserved28 = Value(28.U)
    val Reserved29 = Value(29.U)
    val Reserved30 = Value(30.U)
    val Reserved31 = Value(31.U)
  }

  def modeIsImmediate(mode: Modes.Type): Bool = {
    (mode === Modes.Subi) ||
    (mode === Modes.Addi) ||
    false.B
  }

  class Base(params: AmletParams) extends Instr.Base(params) {
    val mode = Modes()
    val src1 = params.dReg()
    val src2 = params.dReg()
    val predicate = params.pReg()
    val dst = params.bReg()

    def expand(): Expanded = {
      val expanded = Wire(new Expanded(params))
      expanded.mode := mode
      expanded.src1 := src1
      expanded.src2 := src2
      expanded.predicate := predicate
      expanded.dst := dst
      expanded
    }
  }
  
  class Expanded(params: AmletParams) extends Instr.Expanded(params) {
    val mode = Modes()
    val src1 = params.dReg()
    val src2 = params.dReg()
    val predicate = params.pReg()
    val dst = params.bReg()
    
    private val regUtils = new RegUtils(params)

    def getTReads(): Seq[Valid[UInt]] = {
      val src1Read = Wire(Valid(params.tReg()))
      src1Read.valid := true.B
      src1Read.bits := regUtils.dRegToTReg(src1)
      
      val src2Read = Wire(Valid(params.tReg()))
      src2Read.valid := true.B
      src2Read.bits := regUtils.dRegToTReg(src2)
      
      val predicateRead = Wire(Valid(params.tReg()))
      predicateRead.valid := true.B
      predicateRead.bits := regUtils.pRegToTReg(predicate)
      
      Seq(src1Read, src2Read, predicateRead)
    }

    def getTWrites(): Seq[Valid[UInt]] = {
      val dstWrite = Wire(Valid(params.tReg()))
      dstWrite.valid := true.B
      dstWrite.bits := regUtils.bRegToTReg(dst)
      Seq(dstWrite)
    }
  }
  
  class Resolving(params: AmletParams) extends Instr.Resolving(params) {
    val mode = Modes()
    val src1 = new DTaggedSource(params)
    val src2 = new DTaggedSource(params) 
    val predicate = new PTaggedSource(params)
    val old = new BTaggedSource(params)
    val dst = new BTaggedReg(params)

    def isResolved(): Bool = {
      (src1.resolved && src2.resolved && predicate.resolved && predicate.getData) ||
      (old.resolved && predicate.resolved && !predicate.getData)
    }

    def resolve(): Resolved = {
      val resolved = Wire(new Resolved(params))
      resolved.mode := mode
      when (predicate.getData) {
        resolved.src1 := src1.getData
      } .otherwise {
        resolved.src1 := old.getData
      }
      resolved.src2 := src2.getData
      resolved.dst := dst
      resolved.predicate := predicate.getData
      resolved
    }

    def update(writes: ResultBus): Resolving = {
      val resolving = Wire(new Resolving(params))
      resolving.mode := mode
      resolving.src1 := src1.update(writes)
      resolving.src2 := src2.update(writes)
      resolving.old := old.update(writes)
      resolving.predicate := predicate.update(writes)
      resolving.dst := dst
      resolving
    }
  }

  class Resolved(params: AmletParams) extends Instr.Resolved(params) {
    val mode = Modes()
    val src1 = params.dWord()
    val src2 = params.dWord()
    val dst = new BTaggedReg(params)
    val predicate = Bool()
  }

}
